File: Dijkstra.zig
const std = @import("std");

const INF = 1_000_000;

pub fn dijkstra(graph: [][]i32, start: usize) []i32 {
    const n = graph.len;
    var dist = std.heap.page_allocator.alloc(i32, n) catch unreachable;
    var visited = std.heap.page_allocator.alloc(bool, n) catch unreachable;

    for (0..n) |i| {
        dist[i] = INF;
        visited[i] = false;
    }
    dist[start] = 0;

    for (0..n - 1) |_| {
        var min = INF;
        var u: isize = -1;

        for (0..n) |i| {
            if (!visited[i] and dist[i] < min) {
                min = dist[i];
                u = @intCast(i);
            }
        }

        if (u == -1) break;
        visited[@intCast(u)] = true;

        for (0..n) |v| {
            if (!visited[v] and graph[@intCast(u)][v] != 0 and
                dist[@intCast(u)] + graph[@intCast(u)][v] < dist[v])
            {
                dist[v] = dist[@intCast(u)] + graph[@intCast(u)][v];
            }
        }
    }

    return dist;
}

File: Dijkstra.fs
let INF = 1_000_000

let dijkstra (graph:int[,]) start =
    let n = graph.GetLength(0)
    let dist = Array.create n INF
    let visited = Array.create n false

    dist[start] <- 0

    for _ in 0 .. n - 2 do
        let mutable minDist = INF
        let mutable u = -1

        // Trouver le sommet non visité avec la distance minimale
        for i in 0 .. n - 1 do
            if not visited[i] && dist[i] < minDist then
                minDist <- dist[i]
                u <- i

        if u <> -1 then
            visited[u] <- true

            // Mise à jour des distances
            for v in 0 .. n - 1 do
                if not visited[v] && graph[u, v] <> 0 &&
                   dist[u] + graph[u, v] < dist[v] then
                    dist[v] <- dist[u] + graph[u, v]

    dist

[<EntryPoint>]
let main _ =
    let graph = array2D [
        [0; 4; 0; 0; 2]
        [4; 0; 3; 0; 0]
        [0; 3; 0; 5; 0]
        [0; 0; 5; 0; 1]
        [2; 0; 0; 1; 0]
    ]

    let result = dijkstra graph 0
    printfn "%A" result
    0
